# Test Set Evaluation of Top Performing Models

## Libraries

In [28]:
import numpy as np
import matplotlib.pyplot as plt   
%matplotlib inline
from music21 import *
from IPython.display import Audio
from intervaltree import Interval,IntervalTree
import pandas as pd
import seaborn as sns

from sklearn.pipeline import make_pipeline
from sklearn.svm import SVC
import sklearn.model_selection as model_selection
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier

from sklearn.metrics import precision_score, recall_score, f1_score, accuracy_score
from sklearn.metrics import classification_report
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import confusion_matrix
from sklearn.decomposition import PCA
import tensorflow as tf
from sklearn.metrics import confusion_matrix
from sklearn.metrics import plot_confusion_matrix

from collections import Counter
from imblearn.over_sampling import SMOTE
from sklearn.model_selection import train_test_split
import warnings
warnings.simplefilter(action='ignore')




### Import spectral csv features which were extracted from earlier Data PreProcessing notebook

In [2]:
X_train_w = pd.read_csv('../data/X_train_w.csv',index_col = 0)
X_test_w = pd.read_csv('../data/X_test_w.csv',index_col = 0)
y_train_w = pd.read_csv('../data/y_train_w.csv',index_col = 0)
y_test_w = pd.read_csv('../data/y_test_w.csv',index_col = 0)

# Development Results
|       | Model               | Model Variant                                           | Feature Set | Max CV F1 Score | Mean CV F1 Score |
|-------|---------------------|---------------------------------------------------------|-------------|-----------------|------------------|
| 1.1.1 | Logistic Regression | Baseline                                                | Spectral    | 97.6%           | 94.4%            |
| 1.1.2 | Logistic Regression | w/PCA                                                   | Spectral    | 63.6%           | 59.2%            |
| 1.2.1 | Logistic Regression | Baseline                                                | Midi        | 69.6%           | 65.0%            |
| 1.2.2 | Logistic Regression | w/PCA                                                   | Midi        | 31.2%           | 27.4%            |
| 1.3.1 | Logistic Regression | Baseline                                                | Combined    | 91.4%           | 82.4%            |
| 1.3.2 | Logistic Regression | w/PCA                                                   | Combined    | 31.2%           | 29.9%            |
| 2.1.1 | Naive Bayes         | Baseline                                                | Spectral    | 70.9%           | 66.7%            |
| 2.2.1 | Naive Bayes         | Baseline                                                | Midi        | 75.8%           | 58.9%            |
| 2.3.1 | Naive Bayes         | Baseline                                                | Combined    | 76.8%           | 74.3%            |
| 3.1.1 | Decision Tree       | Baseline                                                | Spectral    | 75.7%           | 71.4%            |
| 3.2.1 | Decision Tree       | Baseline                                                | Midi        | 84.8%           | 82.5%            |
| 3.3.1 | Decision Tree       | Baseline                                                | Combined    | 91.8%           | 81.4%            |
| 4.1.1 | Random Forest       | Baseline                                                | Spectral    | 83.9%           | 78.6%            |
| 4.1.2 | Random Forest       | w/SMOTE                                                 | Spectral    | 100%            | 99.6%            |
| 4.2.1 | Random Forest       | Baseline                                                | Midi        | 92.7%           | 87.5%            |
| 4.2.2 | Random Forest       | w/SMOTE                                                 | Midi        | 99.6%           | 99.4%            |
| 4.3.1 | Random Forest       | Baseline                                                | Combined    | 91.3%           | 82.9%            |
| 4.3.2 | Random Forest       | w/SMOTE                                                 | Combined    | 100%            | 99.3%            |
| 5.1.1 | SVM                 | Baseline                                                | Spectral    | 57.6%           | 54.2%            |
| 5.1.2 | SVM                 | w/Hyperparameter Tuning                                 | Spectral    | 94.6%           | 90.4%            |
| 5.1.3 | SVM                 | w/Hyperparameter Tuning +  PCA                          | Spectral    | 91.5%           | 89.1%            |
| 5.1.4 | SVM                 | w/Hyperparameter Tuning +  PCA + SMOTE                  | Spectral    | 100%            | 99.4%            |
| 5.1.5 | SVM                 | w/Hyperparameter Tuning +  PCA + SMOTE + MinMaxScaler   | Spectral    | 75.5%           | 70.2%            |
| 5.1.6 | SVM                 | w/Hyperparameter Tuning +  PCA + SMOTE + StandardScaler | Spectral    | 97.3%           | 96.3%            |
| 5.2.1 | SVM                 | Baseline                                                | Midi        | 65.6%           | 64.1%            |
| 5.3.1 | SVM                 | Baseline                                                | Combined    | 65.6%           | 64.0%            |
| 6.1.1 | Neural Network      | Baseline (ADAM,  5 Hidden Layers)                       | Spectral    |                 | 100%             |
| 6.2.1 | Neural Network      | Baseline (ADAM,  5 Hidden Layers)                       | Midi        |                 | 80.8%            |
| 6.3.1 | Neural Network      | Baseline (ADAM,  5 Hidden Layers)                       | Combined    |                 | 81.2%            |

## Spectral Models by Development Score
|       |               Model |                                          Model Variant | Mean CV F1 Score |
|-------|--------------------:|-------------------------------------------------------:|-----------------:|
| 6.1.1 |      Neural Network |                       Baseline (ADAM, 5 Hidden Layers) |             100% |
| 4.1.2 |       Random Forest |                                                w/SMOTE |           99.60% |
| 5.1.4 |                 SVM |                  w/Hyperparameter Tuning + PCA + SMOTE |           99.40% |
| 5.1.6 |                 SVM | w/Hyperparameter Tuning + PCA + SMOTE + StandardScaler |           96.30% |
| 1.1.1 | Logistic Regression |                                               Baseline |           94.40% |
| 5.1.2 |                 SVM |                                w/Hyperparameter Tuning |           90.40% |
| 5.1.3 |                 SVM |                          w/Hyperparameter Tuning + PCA |           89.10% |
| 4.1.1 |       Random Forest |                                               Baseline |           78.60% |
| 3.1.1 |       Decision Tree |                                               Baseline |           71.40% |
| 5.1.5 |                 SVM |   w/Hyperparameter Tuning + PCA + SMOTE + MinMaxScaler |           70.20% |
| 2.1.1 |         Naive Bayes |                                               Baseline |           66.70% |
| 1.1.2 | Logistic Regression |                                                  w/PCA |           59.20% |
| 5.1.1 |                 SVM |                                               Baseline |           54.20% |

#### Set Random Seed

In [30]:
from numpy.random import seed
seed(1)
from tensorflow.random import set_seed
set_seed(2)

### Neural Network - Baseline

In [32]:
model = tf.keras.Sequential([
    tf.keras.layers.Dense(176, activation='relu'),
    tf.keras.layers.Dense(100, activation='relu'),
    tf.keras.layers.Dense(90, activation='relu'),
    tf.keras.layers.Dense(80, activation='relu'),
    tf.keras.layers.Dense(70, activation='relu'),
    tf.keras.layers.Dense(60, activation='relu'),
    tf.keras.layers.Dense(21)
])
model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])
history = model.fit(X_train_w, y_train_w, epochs=200)
predictions = model.predict(X_test_w)
preds = []
for i in predictions:
    preds.append(np.where(i == i.max())[0][0])
preds
f1 = f1_score(y_test_w, preds, average='weighted')
output_file = open('../output/test_set_output', 'a')
model_no = 'Model #: 1\n'
model_type = 'Neural Network Baseline (wav data only)\n'
model_scores = f'F1 Score: {f1}\n'
output_file.writelines([model_no, model_type, model_scores])
output_file.close()

print('Test Set F1 score = {:.5f}'.format(f1))
print('Confusion Matrix: \n', confusion_matrix(np.array(y_test_w), preds))
ensemble_list = ['Piano Quintet', 'Solo Piano', 'Piano Trio', 'Viola Quintet',
       'String Quartet', 'Clarinet Quintet',
       'Pairs Clarinet-Horn-Bassoon', 'Wind Quintet', 'Accompanied Cello',
       'Accompanied Clarinet', 'Wind and Strings Octet', 'String Sextet',
       'Piano Quartet', 'Horn Piano Trio', 'Solo Violin', 'Solo Flute',
       'Solo Cello', 'Violin and Harpsichord',
       'Clarinet-Cello-Piano Trio', 'Accompanied Violin', 'Wind Octet']
df_cm = pd.DataFrame(confusion_matrix(np.array(y_test_w), preds), index = [i for i in "ABCDEFGHIJK"],
                  columns = [i for i in "ABCDEFGHIJK"])
plt.figure(figsize = (10,7))
sn.heatmap(df_cm, annot=True)

Epoch 1/200
8/8 [==============================] - 1s 2ms/step - loss: 8.3398 - accuracy: 0.2684 
Epoch 2/200
8/8 [==============================] - 0s 4ms/step - loss: 2.3820 - accuracy: 0.4719
Epoch 3/200
8/8 [==============================] - 0s 11ms/step - loss: 1.5722 - accuracy: 0.5714
Epoch 4/200
8/8 [==============================] - 0s 14ms/step - loss: 1.1258 - accuracy: 0.6450
Epoch 5/200
8/8 [==============================] - 0s 4ms/step - loss: 0.9362 - accuracy: 0.7532
Epoch 6/200
8/8 [==============================] - 0s 13ms/step - loss: 0.7898 - accuracy: 0.7576
Epoch 7/200
8/8 [==============================] - 0s 6ms/step - loss: 0.8815 - accuracy: 0.7316
Epoch 8/200
8/8 [==============================] - 0s 8ms/step - loss: 0.8498 - accuracy: 0.7273
Epoch 9/200
8/8 [==============================] - 0s 4ms/step - loss: 0.7027 - accuracy: 0.7749
Epoch 10/200
8/8 [==============================] - 0s 4ms/step - loss: 0.5850 - accuracy: 0.8268
Epoch 11/200
8/8 [=======

8/8 [==============================] - 0s 5ms/step - loss: 0.0765 - accuracy: 0.9784
Epoch 85/200
8/8 [==============================] - 0s 9ms/step - loss: 0.0467 - accuracy: 0.9827
Epoch 86/200
8/8 [==============================] - 0s 5ms/step - loss: 0.0684 - accuracy: 0.9697
Epoch 87/200
8/8 [==============================] - 0s 5ms/step - loss: 0.1808 - accuracy: 0.9610
Epoch 88/200
8/8 [==============================] - 0s 6ms/step - loss: 0.1300 - accuracy: 0.9610
Epoch 89/200
8/8 [==============================] - 0s 6ms/step - loss: 0.0480 - accuracy: 0.9784
Epoch 90/200
8/8 [==============================] - 0s 9ms/step - loss: 0.0451 - accuracy: 0.9827
Epoch 91/200
8/8 [==============================] - 0s 5ms/step - loss: 0.0239 - accuracy: 0.9957
Epoch 92/200
8/8 [==============================] - 0s 5ms/step - loss: 0.0260 - accuracy: 0.9913
Epoch 93/200
8/8 [==============================] - 0s 3ms/step - loss: 0.0138 - accuracy: 0.9957
Epoch 94/200
8/8 [===============

8/8 [==============================] - 0s 5ms/step - loss: 4.8236e-04 - accuracy: 1.0000
Epoch 166/200
8/8 [==============================] - 0s 6ms/step - loss: 4.6637e-04 - accuracy: 1.0000
Epoch 167/200
8/8 [==============================] - 0s 6ms/step - loss: 4.6418e-04 - accuracy: 1.0000
Epoch 168/200
8/8 [==============================] - 0s 5ms/step - loss: 4.4714e-04 - accuracy: 1.0000
Epoch 169/200
8/8 [==============================] - 0s 10ms/step - loss: 4.5658e-04 - accuracy: 1.0000
Epoch 170/200
8/8 [==============================] - 0s 4ms/step - loss: 4.7675e-04 - accuracy: 1.0000
Epoch 171/200
8/8 [==============================] - 0s 5ms/step - loss: 5.1564e-04 - accuracy: 1.0000
Epoch 172/200
8/8 [==============================] - 0s 5ms/step - loss: 4.6901e-04 - accuracy: 1.0000
Epoch 173/200
8/8 [==============================] - 0s 4ms/step - loss: 4.2640e-04 - accuracy: 1.0000
Epoch 174/200
8/8 [==============================] - 0s 9ms/step - loss: 4.0539e-04 - 

### Random Forest - w/SMOTE

In [26]:
# todo: I think we need to re-import the data for this cell, if this is run multiple times then y_train_w gets overwritten
y_train_w = pd.read_csv('../data/y_train_w.csv',index_col = 0)

#Drop any ensemble types with counts below 6. This because Expected n_neighbors <= n_samples,  
#but n_samples = 3, n_neighbors = 6
counts = y_train_w['ensemble'].value_counts()
X_train_w = X_train_w[~y_train_w['ensemble'].isin(counts[counts < 6].index)]
y_train_w = y_train_w[~y_train_w['ensemble'].isin(counts[counts < 6].index)]

# Fit and apply the transform
X_SMOTE_w, y_SMOTE_w = SMOTE(random_state=101).fit_resample(X_train_w, y_train_w)

from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(random_state=1)
rf.fit(X_SMOTE_w, y_SMOTE_w)
predictions = rf.predict(X_test_w)
f1 = f1_score(y_test_w, predictions, average='weighted')
output_file = open('../output/test_set_output', 'a')
model_no = 'Model #: 2\n'
model_type = 'Random Forest (wav data only)\n'
model_scores = f'F1 Score: {f1}\n'
output_file.writelines([model_no, model_type, model_scores])
output_file.close()

print('Test Set F1 score = {:.2%}'.format(f1))

Test Set F1 score = 69.14%


### SVM - w/Hyperparameter Tuning + PCA + SMOTE

In [5]:
#Drop any ensemble types with counts below 6
counts = y_train_w['ensemble'].value_counts()
X_train_w_smote = X_train_w[~y_train_w['ensemble'].isin(counts[counts < 6].index)]
y_train_w_smote = y_train_w[~y_train_w['ensemble'].isin(counts[counts < 6].index)]

# fit and apply the transform
X_SMOTE, y_SMOTE = SMOTE().fit_resample(X_train_w_smote, y_train_w_smote.values.ravel())

#Perform PCA with 15 components
pca=PCA(n_components = 15)
pca.fit(X_SMOTE)
train_pca_w_smote = pca.transform(X_SMOTE)
X_test_w_pca = pca.transform(X_test_w)


from sklearn.model_selection import GridSearchCV
  
# defining parameter range
param_grid = {'C': np.logspace(-4, 4, 20), 
              'gamma': np.logspace(-3, 2, 6),
              'kernel': ['linear','rbf','poly']} 
  
grid = GridSearchCV(SVC(), param_grid , cv=5)
  
# fitting the model for grid search
grid.fit(train_pca_w_smote, y_SMOTE)

#Create SVM model with best hyperparameters, PCA and SMOTE

svm = make_pipeline(SVC(kernel=grid.best_params_['kernel'], C = grid.best_params_['C'] , gamma=grid.best_params_['gamma']))
svm.fit(train_pca_w_smote, y_SMOTE)
predictions = svm.predict(X_test_w_pca)
f1 = f1_score(y_test_w, predictions, average='weighted')
output_file = open('../output/test_set_output', 'a')
model_no = 'Model #: 3\n'
model_type = 'SVM - Spectral - w/Hyperparameter Tuning + PCA + SMOTE (wav data only)\n'
model_scores = f'F1 Score: {f1}\n'
output_file.writelines([model_no, model_type, model_scores])
output_file.close()

print("Test Set F1-score: {:.2%}".format(f1))


Test Set F1-score: 72.81%


### SVM - w/Hyperparameter Tuning + PCA + SMOTE + StandardScaler	

In [6]:
#### Standard Scaler Application

#Perform PCA with 15 components
pca=PCA(n_components = 15)
pca.fit(X_SMOTE)
train_pca_w_smote = pca.transform(X_SMOTE)
X_test_w_pca = pca.transform(X_test_w)


#Create SVM model with best hyperparameters, PCA and SMOTE and StandardScaler

svm = make_pipeline(StandardScaler() ,SVC(kernel=grid.best_params_['kernel'], C = grid.best_params_['C'] , gamma=grid.best_params_['gamma']))
svm.fit(train_pca_w_smote, y_SMOTE)


predictions = svm.predict(X_test_w_pca)
f1 = f1_score(y_test_w, predictions,average='weighted')
output_file = open('../output/test_set_output', 'a')
model_no = 'Model #: 4\n'
model_type = 'SVM - Spectral - w/Hyperparameter Tuning + PCA + SMOTE + StandardScaler (wav data only)\n'
model_scores = f'F1 Score: {f1}\n'
output_file.writelines([model_no, model_type, model_scores])
output_file.close()
print("Test Set F1-score: {:.2%}".format(f1))


Test Set F1-score: 68.41%


### Logistic Regression - Baseline

In [7]:
#Create Baseline Logistic Model
lr = LogisticRegression(C = 0.5, solver='liblinear', multi_class='auto')
lr.fit(np.array(X_train_w), np.array(y_train_w))
predictions = lr.predict(X_test_w)
f1 = f1_score(y_test_w, predictions,average='weighted')
output_file = open('../output/test_set_output', 'a')
model_no = 'Model #: 5\n'
model_type = 'Logistic Regression - Spectral - Baseline (wav data only)\n'
model_scores = f'F1 Score: {f1}\n'
output_file.writelines([model_no, model_type, model_scores])
output_file.close()
print("Test Set F1-score: {:.2%}".format(f1))


Test Set F1-score: 72.76%


### SVM - Spectral - w/Hyperparameter Tuning

In [8]:
#Find best hyperparameters
param_grid = {'C': np.logspace(-4, 4, 20), 
              'gamma': np.logspace(-3, 2, 6),                     
              'kernel': ['linear','rbf','poly']} 
  
grid = GridSearchCV(SVC(), param_grid, cv=5)
  
# fitting the model for grid search
grid.fit(X_train_w, y_train_w.values.ravel())

#Create Model with best hyperparameters

svm = make_pipeline(SVC(kernel=grid.best_params_['kernel'], C = grid.best_params_['C'] , gamma=grid.best_params_['gamma']))
svm.fit(X_train_w, y_train_w.values.ravel())
predictions = svm.predict(X_test_w)
f1 = f1_score(y_test_w, predictions,average='weighted')
output_file = open('../output/test_set_output', 'a')
model_no = 'Model #: 6\n'
model_type = 'SVM - Spectral - w/Hyperparameter Tuning (wav data only)\n'
model_scores = f'F1 Score: {f1}\n'
output_file.writelines([model_no, model_type, model_scores])
output_file.close()
print("Test Set F1-score: {:.2%}".format(f1))

Test Set F1-score: 70.17%


### SVM - Spectral - W/Hyperparameter Tuning + PCA

In [9]:
#Perform PCA with 15 components
pca=PCA(n_components = 15)
pca.fit(X_train_w)
train_pca_w = pca.transform(X_train_w)
X_test_w_pca = pca.transform(X_test_w)

from sklearn.model_selection import GridSearchCV
  
# defining parameter range
param_grid = {'C': np.logspace(-4, 4, 20), 
              'gamma': np.logspace(-3, 2, 6),
              'kernel': ['linear','rbf','poly']} 
  
grid = GridSearchCV(SVC(), param_grid , cv=5)
  
# fitting the model for grid search
grid.fit(train_pca_w, y_train_w.values.ravel())

#Create Model with best hyperparameters and PCA
svm = make_pipeline(SVC(kernel=grid.best_params_['kernel'], C = grid.best_params_['C'] , gamma=grid.best_params_['gamma']))
svm.fit(train_pca_w, y_train_w.values.ravel())
predictions = svm.predict(X_test_w_pca)
f1 = f1_score(y_test_w, predictions,average='weighted')
output_file = open('../output/test_set_output', 'a')
model_no = 'Model #: 7\n'
model_type = 'SVM - Spectral - W/Hyperparameter Tuning + PCA (wav data only)\n'
model_scores = f'F1 Score: {f1}\n'
output_file.writelines([model_no, model_type, model_scores])
output_file.close()
print("Test Set F1-score: {:.2%}".format(f1))


{'C': 0.0018329807108324356, 'gamma': 0.001, 'kernel': 'linear'}
Test Set F1-score: 66.30%


### Random Forest - Baseline

In [10]:
rf = RandomForestClassifier()
rf.fit(X_train_w, y_train_w)
predictions = rf.predict(X_test_w)
f1 = f1_score(y_test_w, predictions,average='weighted')
output_file = open('../output/test_set_output', 'a')
model_no = 'Model #: 8\n'
model_type = 'Random Forest - Midi - Baseline (wav data only)\n'
model_scores = f'F1 Score: {f1}\n'
output_file.writelines([model_no, model_type, model_scores])
output_file.close()
print("Test Set F1-score: {:.2%}".format(f1))

Test Set F1-score: 64.01%


### Decision Tree - Baseline

In [11]:
dt = DecisionTreeClassifier()
dt.fit(X_train_w, y_train_w)
predictions = dt.predict(X_test_w)
f1 = f1_score(y_test_w, predictions, average = 'weighted')
output_file = open('../output/test_set_output', 'a')
model_no = 'Model #: 9\n'
model_type = 'Decision Tree Baseline Model (wav data only)\n'
model_scores = f'F1 Score: {f1}\n'
output_file.writelines([model_no, model_type, model_scores])
output_file.close()
print("Test Set F1-score: {:.2%}".format(f1))

Test Set F1-score: 68.29%


### SVM - w/Hyperparameter Tuning + PCA + SMOTE + MinMaxScaler

In [12]:
#### MinMaxScaler Application

#Perform PCA with 15 components
pca=PCA(n_components = 15)
pca.fit(X_SMOTE)
train_pca_w_smote = pca.transform(X_SMOTE)
X_test_w_pca = pca.transform(X_test_w)


#Create SVM model with best hyperparameters, PCA and SMOTE and MinMaxScaler

svm = make_pipeline(MinMaxScaler() ,SVC(kernel=grid.best_params_['kernel'], C = grid.best_params_['C'] , gamma=grid.best_params_['gamma']))
svm.fit(train_pca_w_smote, y_SMOTE)
predictions = svm.predict(X_test_w_pca)
f1 = f1_score(y_test_w, predictions, average = 'weighted')

#Apply 5 fold cross validation scoring to make sure results are reliable
# Print and save output
output_file = open('../output/test_set_output', 'a')
model_no = 'Model #: 10\n'
model_type = 'SVM with Hyperparameter tuning, PCA, SMOTE, and Min Max Scaler (wav data only)\n'
model_scores = f'F1 Score: {f1}\n'
output_file.writelines([model_no, model_type, model_scores])
output_file.close()
print("Test Set F1-score: {:.2%}".format(f1))

Test Set F1-score: 34.85%


### Naive Bayes - Baseline

In [13]:
gnb = GaussianNB()
gnb.fit(X_train_w,y_train_w)
predictions = gnb.predict(X_test_w)
f1 = f1_score(y_test_w, predictions, average = 'weighted')

output_file = open('../output/test_set_output', 'a')
model_no = 'Model #: 11\n'
model_type = 'Naive Bayes Baseline Model (wav data only)\n'
model_scores = f'F1 Score: {f1}\n'
output_file.writelines([model_no, model_type, model_scores])
output_file.close()
print("Test Set F1-score: {:.2%}".format(f1))

Test Set F1-score: 60.57%


### Logistic Regression - w/PCA

In [14]:
pca = PCA(n_components = 5)
pca.fit(X_train_w)
X_train_pca = pca.transform(X_train_w)
X_test_w_pca = pca.transform(X_test_w)

lr = LogisticRegression(C = 0.5, max_iter = 100,solver='liblinear', multi_class='auto')
lr.fit(X_train_pca,y_train_w)
predictions = lr.predict(X_test_w_pca)
f1 = f1_score(y_test_w, predictions, average = 'weighted')

output_file = open('../output/test_set_output', 'a')
model_no = 'Model #: 12\n'
model_type = 'Logistic Regression with PCA (wav data only)\n'
model_scores = f'F1 Score: {f1}\n'
output_file.writelines([model_no, model_type, model_scores])
output_file.close()
print("Test Set F1-score: {:.2%}".format(f1))

Test Set F1-score: 54.79%


### SVM - Baseline

In [15]:
#Create baseline SVM model with Spectral Data

svm = make_pipeline(SVC())
svm.fit(X_train_w, y_train_w.values.ravel())
predictions = svm.predict(X_test_w)
f1 = f1_score(y_test_w, predictions, average = 'weighted')

# prints and saves output
output_file = open('../output/test_set_output', 'a')
model_no = 'Model #: 13\n'
model_type = 'SVM Baseline Model (wav data only)\n'
model_scores = f'F1 Score: {f1}\n'
output_file.writelines([model_no, model_type, model_scores])
output_file.close()
print("Test Set F1-score: {:.2%}".format(f1))

Test Set F1-score: 47.05%


## Final Test Results
|       |               Model |                                          Model Variant | Mean CV F1 Score | Test Set F1 Score |
|-------|--------------------:|-------------------------------------------------------:|-----------------:|-------------------|
| 6.1.1 |      Neural Network |                       Baseline (ADAM, 5 Hidden Layers) |             100% | 73.1%             |
| 4.1.2 |       Random Forest |                                                w/SMOTE |           99.60% | 69.14%            |
| 5.1.4 |                 SVM |                  w/Hyperparameter Tuning + PCA + SMOTE |           99.40% | 72.81%            |
| 5.1.6 |                 SVM | w/Hyperparameter Tuning + PCA + SMOTE + StandardScaler |           96.30% | 68.41%            |
| 1.1.1 | Logistic Regression |                                               Baseline |           94.40% | 72.76%            |
| 5.1.2 |                 SVM |                                w/Hyperparameter Tuning |           90.40% | 70.17%            |
| 5.1.3 |                 SVM |                          w/Hyperparameter Tuning + PCA |           89.10% | 66.30%            |
| 4.1.1 |       Random Forest |                                               Baseline |           78.60% | 64.01%            |
| 3.1.1 |       Decision Tree |                                               Baseline |           71.40% | 68.29%            |
| 5.1.5 |                 SVM |   w/Hyperparameter Tuning + PCA + SMOTE + MinMaxScaler |           70.20% | 34.85%            |
| 2.1.1 |         Naive Bayes |                                               Baseline |           66.70% | 60.57%            |
| 1.1.2 | Logistic Regression |                                                  w/PCA |           59.20% | 54.79%            |
| 5.1.1 |                 SVM |                                               Baseline |           54.20% | 47.05%            |